In [119]:
#reading multiple csv files
import pandas as pd
import glob #glob is amazing

csv_files = glob.glob(r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\*.csv") #reads any file in the directory that ends with .csv

datasets = {}

for file in csv_files: 
    data = pd.read_csv(file) #read each file one by one
    datasets[file] = data #put each file into the dictionary so we can call those files without having to read each one again



In [120]:
#inspecting customers dataset

customers = datasets[r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_customers_dataset.csv"]

customers.head(10)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP
5,879864dab9bc3047522c92c82e1212b8,4c93744516667ad3b8f1fb645a3116a4,89254,jaragua do sul,SC
6,fd826e7cf63160e536e0908c76c3f441,addec96d2e059c80c30fe6871d30d177,4534,sao paulo,SP
7,5e274e7a0c3809e14aba7ad5aae0d407,57b2a98a409812fe9618067b6b8ebe4f,35182,timoteo,MG
8,5adf08e34b2e993982a47070956c5c65,1175e95fb47ddff9de6b2b06188f7e0d,81560,curitiba,PR
9,4b7139f34592b3a31687243a302fa75b,9afe194fb833f79e300e37e580171f22,30575,belo horizonte,MG


**Customer id, customer_unique_id** look like an idetifier. not human readable.

Questions: 
- "is it unique?"
- "does it appear in other tables?"
- "does it have a consistent format?"
- "are there missing values?"
- "are there duplicates?"
- "why would a table need 2 ids?:

**Customer_zip_code_prefix** - in the data head itself i notice two values 1151 and 4534 having fewer than 5 digits. 

lets inpect another table 

In [121]:
#inspectin orders_dataset

orders = datasets[r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_orders_dataset.csv"]

orders.head(10)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00
5,a4591c265e18cb1dcee52889e2d8acc3,503740e9ca751ccdda7ba28e9ab8f608,delivered,2017-07-09 21:57:05,2017-07-09 22:10:13,2017-07-11 14:58:04,2017-07-26 10:57:55,2017-08-01 00:00:00
6,136cce7faa42fdb2cefd53fdc79a6098,ed0271e0b7da060a393796590e7b737a,invoiced,2017-04-11 12:22:08,2017-04-13 13:25:17,NaN,NaN,2017-05-09 00:00:00
7,6514b8ad8028c9f2cc2374ded245783f,9bdf08b4b3b52b5526ff42d37d47f222,delivered,2017-05-16 13:10:30,2017-05-16 13:22:11,2017-05-22 10:07:46,2017-05-26 12:55:51,2017-06-07 00:00:00
8,76c6e866289321a7c93b82b54852dc33,f54a9f0e6b351c431402b8461ea51999,delivered,2017-01-23 18:29:09,2017-01-25 02:50:47,2017-01-26 14:16:31,2017-02-02 14:08:10,2017-03-06 00:00:00
9,e69bfb5eb88e0ed6a785585b27e16dbf,31ad1d1b63eb9962463f764d4e6e0c9d,delivered,2017-07-29 11:55:02,2017-07-29 12:05:32,2017-08-10 19:45:24,2017-08-16 17:14:30,2017-08-23 00:00:00


**order_status** is important, it describes the state of an order
the date columns need to be examined for consistency and datatype
NaN values need to be quantified, not just noticed visually
the different date columns probably have difference business meaning so we shouldnt assume they all behave identically

In [122]:
orders.info()
orders.isnull().sum()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 6.1 MB


order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

**IN ORDERS DATASET**      

- `customer_id` appears in both the customers and orders tables.
- `order_id` is present in the orders table.
- Several timestamp/date columns are present.
- Some date columns contain missing values.
- All columns are currently stored as strings.


**MISSING VALUES:**

`order_approved_at` = 99441 - 99281 = 160

`order_delivered_carrirer_date` 99441 - 97658 =  1783

`order_delivered_customer_date` 99441 - 96476 = 2965


**Questions:**

- "How do we fill these NaNs?"
- "Why are these values missing?"
- "Are missing delivery dates associated with "order status" values?"
- "Are all timestamp columns consistently formatted?"

if an order hasnt been delivered, it might naturally have no value.

In [123]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

### Missing delivery date investigation

- `order_status` contains 96,478 `delivered` orders out of 99,441 total orders.
- `order_delivered_customer_date` has 2,965 missing values.
- There are 2,963 non-delivered orders:
  `99,441 - 96,478 = 2,963`
- Therefore, 2 missing `order_delivered_customer_date` values appear to belong to orders marked as `delivered`.

### Question

- Why do these 2 orders have `order_status = delivered` but no `order_delivered_customer_date`?
- Need to inspect the 2 records before deciding whether they are data-quality errors.

In [124]:
orders[orders["order_delivered_customer_date"].isnull() & (orders["order_status"] == "delivered")]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaN,2017-12-18 00:00:00
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaN,2018-07-16 00:00:00
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaN,2018-07-24 00:00:00
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaN,NaN,2017-06-23 00:00:00
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaN,2018-06-26 00:00:00
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaN,2018-07-19 00:00:00


In [125]:
pd.crosstab(
    orders["order_status"],
    orders["order_delivered_customer_date"].isnull()
)

order_delivered_customer_date,False,True
order_status,,
approved,0,2
canceled,6,619
created,0,5
delivered,96470,8
invoiced,0,314
processing,0,301
shipped,0,1107
unavailable,0,609


### Missing delivery date investigation

- `order_delivered_customer_date` has 2,965 missing values.
- Filtering for orders where:
  - `order_status == "delivered"`
  - `order_delivered_customer_date` is missing
- returned 8 records.
- Therefore, missing delivery dates are not exclusively associated with non-delivered orders.

### Question

- Why are these 8 orders marked as `delivered` despite having no customer delivery date?
- Need to investigate these records and determine whether they represent legitimate missing data or data-quality issues.

In [126]:
orders["order_purchase_timestamp"].tail(20)


99421    2017-07-05 15:55:19
99422    2018-05-25 17:05:17
99423    2017-10-12 20:54:11
99424    2017-11-21 20:18:03
99425    2018-03-21 19:47:18
99426    2017-05-15 09:46:26
99427    2018-05-12 13:42:24
99428    2018-06-22 20:53:29
99429    2017-11-05 16:17:53
99430    2018-03-04 22:48:38
99431    2017-10-27 15:21:00
99432    2017-12-20 09:52:41
99433    2017-10-04 19:57:37
99434    2017-01-27 00:30:03
99435    2017-02-23 09:05:12
99436    2017-03-09 09:54:05
99437    2018-02-06 12:58:58
99438    2017-08-27 14:46:43
99439    2018-01-08 21:28:27
99440    2018-03-08 20:57:30
Name: order_purchase_timestamp, dtype: str

In [127]:
pd.to_datetime(orders["order_purchase_timestamp"], errors = 'coerce')
orders["order_purchase_timestamp"].isnull().sum()

np.int64(0)

### `order_purchase_timestamp`

- Stored as `str` in the raw dataset.
- Values follow the apparent format `YYYY-MM-DD HH:MM:SS`.
- `pd.to_datetime(..., errors="coerce")` successfully parsed all 99,441 values.
- 0 values became `NaT`.
- No invalid/unparseable values detected in this column.

In [128]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

In [129]:
date_columns = ['order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date']

for column in date_columns:
    converted = pd.to_datetime(orders[column], errors = "coerce")
    print(column, converted.isnull().sum())



order_purchase_timestamp 0
order_approved_at 160
order_delivered_carrier_date 1783
order_delivered_customer_date 2965
order_estimated_delivery_date 0


### Date columns

All date columns are stored as `str` in the raw dataset.

After testing conversion with `pd.to_datetime(..., errors="coerce")`:

| Column | Failed/NaT values |
|---|---|
| `order_purchase_timestamp` | 0 |
| `order_approved_at` | 160 |
| `order_delivered_carrier_date` | 1,783 |
| `order_delivered_customer_date` | 2,965 |
| `order_estimated_delivery_date` | 0 |

Observations:
- No invalid date strings were detected.
- The missing values in the three affected columns remain missing after conversion.
- Date values appear to follow a consistent `YYYY-MM-DD HH:MM:SS` format.
- These columns can likely be converted to datetime during the cleaning phase.

In [130]:
orders.order_id.nunique()

99441

### `order_id`

- Total rows: 99,441
- Unique `order_id` values: 99,441
- Every row has a unique `order_id`.
- `order_id` appears to function as the unique identifier for each order.

In [131]:
orders["customer_id"].isin(customers["customer_id"]).value_counts()

customer_id
True    99441
Name: count, dtype: int64

### Referential Integrity

Checks whether every `customer_id` in `orders` exists in `customers`.

```python
orders["customer_id"].isin(customers["customer_id"]).value_counts()

safe to assume customer.customer_id = orders.customer_id

In [132]:
customers["customer_unique_id"].value_counts().head(10)

customer_unique_id
8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
6469f99c1f9dfae7733b25662e7f1782     7
ca77025e7201e3b30c44b472ff346268     7
47c1a3033b8b77b3ab6e109eb4d5fdf3     6
dc813062e0fc23409cd255f7f53c7074     6
63cfc61cee11cbe306bff5857d00bfe4     6
f0e310a6839dce9de1638e0fe5ab282a     6
de34b16117594161a6a89c50b289d35a     6
Name: count, dtype: int64

In [133]:
customers[customers["customer_unique_id"] == "8d50f5eadf50201ccdcedfb9e2ac8455"]

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
14186,1bd3585471932167ab72a84955ebefea,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
15321,a8fabc805e9a10a3c93ae5bff642b86b,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
16654,897b7f72042714efaa64ac306ba0cafc,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
36122,b2b13de0770e06de50080fea77c459e6,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
38073,42dbc1ad9d560637c9c4c1533746f86d,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
40141,dfb941d6f7b02f57a44c3b7c3fefb44b,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
48614,65f9db9dd07a4e79b625effa4c868fcb,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
52574,1c62b48fb34ee043310dcb233caabd2e,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
58707,a682769c4bc10fc6ef2101337a6c83c9,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP
67996,6289b75219d757a56c0cce8d9e427900,8d50f5eadf50201ccdcedfb9e2ac8455,4045,sao paulo,SP


In [134]:
customer = customers [customers["customer_unique_id"] == customers["customer_unique_id"].value_counts().index[0]] 

customer

customer_orders = orders[orders["customer_id"].isin(customer["customer_id"])

]

customer_orders

len(customer)


17

In [135]:
customers.groupby("customer_unique_id")["customer_id"].nunique().value_counts().sort_index() #group all customer_ids to customer_unique_id, count their unique values, and sort them

customer_id
1     93099
2      2745
3       203
4        30
5         8
6         6
7         3
9         1
17        1
Name: count, dtype: int64

### Inspection Flow
customer_id
    ↓

Is it unique?
    ↓

Yes, 99,441 unique IDs
    ↓

customer_unique_id
    ↓

Is IT unique?
    ↓

No, 96,096 unique IDs
    ↓
    
So some customer_unique_id values repeat
    ↓ 

How much do they repeat?
    ↓

value_counts()
    ↓

One customer appears 17 times
    ↓

Is this just one weird customer?
    ↓

We need to examine ALL customers
    ↓
    
groupby()

In [136]:
orders.groupby("customer_id")["order_id"].nunique().value_counts().sort_index()

order_id
1    99441
Name: count, dtype: int64

### Customer ID relationship

- Every `customer_id` in `orders` is associated with exactly one `order_id`.
- `customer_unique_id` can appear across multiple `customer_id` records.
- This indicates that multiple orders can belong to the same underlying customer.
- No cleaning is required for these repeated `customer_unique_id` values.

In [137]:
customers.groupby("customer_zip_code_prefix")["customer_state"].nunique().value_counts().sort_index()

customer_state
1    14994
Name: count, dtype: int64

### Geographic consistency

- Checked the relationship between `customer_zip_code_prefix` and `customer_state`.
- All 14,994 ZIP-code prefixes are associated with exactly one state.
- No ZIP/state conflicts were found.

In [138]:
customers.groupby("customer_zip_code_prefix")["customer_city"].nunique().value_counts().sort_index()

customer_city
1    14955
2       38
3        1
Name: count, dtype: int64

In [139]:
customers.groupby("customer_zip_code_prefix")["customer_state"].nunique().value_counts()

customer_state
1    14994
Name: count, dtype: int64

In [140]:
customers.groupby("customer_zip_code_prefix")["customer_city"].nunique().value_counts()

customer_city
1    14955
2       38
3        1
Name: count, dtype: int64

In [141]:
conflicting_zips = (customers.groupby("customer_zip_code_prefix")["customer_city"].nunique())
conflicting_zips = conflicting_zips[conflicting_zips>1].index

customers[customers["customer_zip_code_prefix"].isin(conflicting_zips)][["customer_zip_code_prefix", "customer_city", 'customer_state']]

,customer_zip_code_prefix,customer_city,customer_state
54,23970,parati,RJ
60,72270,brasilia,DF
142,6835,embu,SP
633,6824,embu das artes,SP
695,85139,vitoria,PR
...,...,...,...
98831,15720,palmeira d'oeste,SP
98975,6810,embu das artes,SP
99135,6806,embu,SP
99367,73060,brasilia,DF


## Conclusion

The initial raw data inspection is complete for the `customers` and `orders` datasets.

Key findings:
- Both datasets contain the expected columns and records.
- No missing values were found in `customers`.
- Missing values were found in several `orders` date columns.
- All order purchase timestamps were successfully parsed as dates.
- `customer_id` values are consistent between `customers` and `orders`.
- Each `customer_id` is associated with one order in this dataset.
- `customer_unique_id` reveals that some customers placed multiple orders.
- Some ZIP codes contain multiple city names, which may represent naming inconsistencies and will be considered during cleaning.

The remaining Olist datasets still need to be inspected before the cleaning phase begins.

Next step: continue the raw data inspection with `order_items`.

In [142]:
order_items = datasets[r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_order_items_dataset.csv"]
order_items


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14
...,...,...,...,...,...,...,...
112645,fffc94f6ce00a00581880bf54a75a037,1,4aa6014eceb682077f9dc4bffebc05b0,b8bc237ba3788b23da09c0f1f3a3288c,2018-05-02 04:11:01,299.99,43.41
112646,fffcd46ef2263f404302a634eb57f7eb,1,32e07fd915822b0765e448c4dd74c828,f3c38ab652836d21de61fb8314b69182,2018-07-20 04:31:48,350.00,36.53
112647,fffce4705a9662cd70adb13d4a31832d,1,72a30483855e2eafc67aee5dc2560482,c3cfdc648177fdbbbb35635a37472c53,2017-10-30 17:14:25,99.90,16.95
112648,fffe18544ffabc95dfada21779c9644f,1,9c422a519119dcad7575db5af1ba540e,2b3e4a2a3ea8e01938cabda2a3e5cc79,2017-08-21 00:04:32,55.99,8.72


### Columns and Data Types

- 7 columns were found.
- `shipping_limit_date` is stored as a string and will be considered for date conversion during cleaning.
- `price` and `freight_value` are stored as `float64`.
- `order_item_id` is stored as `int64`.
- ID columns are stored as strings.

In [143]:
order_items


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14
...,...,...,...,...,...,...,...
112645,fffc94f6ce00a00581880bf54a75a037,1,4aa6014eceb682077f9dc4bffebc05b0,b8bc237ba3788b23da09c0f1f3a3288c,2018-05-02 04:11:01,299.99,43.41
112646,fffcd46ef2263f404302a634eb57f7eb,1,32e07fd915822b0765e448c4dd74c828,f3c38ab652836d21de61fb8314b69182,2018-07-20 04:31:48,350.00,36.53
112647,fffce4705a9662cd70adb13d4a31832d,1,72a30483855e2eafc67aee5dc2560482,c3cfdc648177fdbbbb35635a37472c53,2017-10-30 17:14:25,99.90,16.95
112648,fffe18544ffabc95dfada21779c9644f,1,9c422a519119dcad7575db5af1ba540e,2b3e4a2a3ea8e01938cabda2a3e5cc79,2017-08-21 00:04:32,55.99,8.72


### Observations 
112650 entries


In [144]:
order_items["order_id"].value_counts()

order_id
8272b63d03f5f79c56e9e4120aec44ef    21
1b15974a0141d54e36626dca3fdc731a    20
ab14fdcfbe524636d65ee38360e22ce8    20
428a2f660dc84138d969ccd69a0ab6d5    15
9ef13efd6949e4573a18964dd1bbe7f5    15
                                    ..
fffc94f6ce00a00581880bf54a75a037     1
fffcd46ef2263f404302a634eb57f7eb     1
fffce4705a9662cd70adb13d4a31832d     1
fffe18544ffabc95dfada21779c9644f     1
fffe41c64501cc87c801fd61db3f6244     1
Name: count, Length: 98666, dtype: int64

In [145]:
order_items[order_items["order_id"] == "8272b63d03f5f79c56e9e4120aec44ef"][["order_id", "order_item_id"]]

,order_id,order_item_id
57297,8272b63d03f5f79c56e9e4120aec44ef,1
57298,8272b63d03f5f79c56e9e4120aec44ef,2
57299,8272b63d03f5f79c56e9e4120aec44ef,3
57300,8272b63d03f5f79c56e9e4120aec44ef,4
57301,8272b63d03f5f79c56e9e4120aec44ef,5
57302,8272b63d03f5f79c56e9e4120aec44ef,6
57303,8272b63d03f5f79c56e9e4120aec44ef,7
57304,8272b63d03f5f79c56e9e4120aec44ef,8
57305,8272b63d03f5f79c56e9e4120aec44ef,9
57306,8272b63d03f5f79c56e9e4120aec44ef,10


### Observation

`order_item_id` shows incremental behavior within each `order_id`.

For example, an order with multiple items has `order_item_id` values progressing from 1, 2, 3, and so on.

This suggests that `order_item_id` represents the item's position within an order rather than a globally unique identifier.

In [146]:
order_range = order_items.groupby("order_id")["order_item_id"].agg(["count", "max"])
order_range[order_range["count"] > 1]

,count,max
order_id,,
0008288aa423d2a3f00fcb17cd7d8719,2,2
00143d0f86d6fbd9f9b38ab440ac16f5,3,3
001ab0a7578dd66cd4b0a71f5b6e1e41,3,3
001d8f0e34a38c37f7dba2a37d4eba8b,2,2
002c9def9c9b951b1bec6d50753c9891,2,2
...,...,...
ffd84ab39cd5e873d8dba24342e65c01,2,2
ffe4b41e99d39f0b837a239110260530,2,2
ffecd5a79a0084f6a592288c67e3c298,3,3


In [147]:
order_range = order_items.groupby("order_id")["order_item_id"].agg(["count", "max"])

order_range[order_range["count"] != order_range["max"]].head(40)

,count,max
order_id,,


### Order Item Sequence

- `order_item_id` is sequential within each `order_id`.
- The combination of `order_id` and `order_item_id` contains no duplicates.
- For every order, the number of item rows matches the maximum `order_item_id`.
- No sequence inconsistencies were found.

In [148]:
order_items["order_id"].isin(orders["order_id"]).value_counts()

order_id
True    112650
Name: count, dtype: int64

In [149]:
products = datasets[r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_products_dataset.csv"]

order_items["product_id"].isin(products["product_id"]).value_counts()

product_id
True    112650
Name: count, dtype: int64

## Order Items

### Observations

- 112,650 rows and 7 columns.
- No missing values were found.
- `shipping_limit_date` is stored as a string and should be considered for date conversion during cleaning.
- `order_id` is not unique because one order can contain multiple items.
- `order_item_id` behaves sequentially within each `order_id`.
- No duplicate `order_id` + `order_item_id` combinations were found.
- All `order_id` values exist in the `orders` dataset.
- All `product_id` values exist in the `products` dataset.

In [150]:
order_payments = datasets[r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_order_payments_dataset.csv"]

order_payments.shape

(103886, 5)

In [151]:
order_payments.info()

<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 4.0 MB


In [152]:
order_payments["order_id"].isin(orders["order_id"]).value_counts()

order_id
True    103886
Name: count, dtype: int64

In [153]:
order_payments.head(10)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45
5,298fcdf1f73eb413e4d26d01b25bc1cd,1,credit_card,2,96.12
6,771ee386b001f06208a7419e4fc1bbd7,1,credit_card,1,81.16
7,3d7239c394a212faae122962df514ac7,1,credit_card,3,51.84
8,1f78449c87a54faf9e96e88ba1491fa9,1,credit_card,6,341.09
9,0573b5e23cbd798006520e1d5b4c6714,1,boleto,1,51.95


In [154]:
payment_sequence = order_payments.groupby("order_id")["payment_sequential"].agg(["count", "max"])

payment_sequence[(payment_sequence["count"] ==  payment_sequence["max"]) & (payment_sequence["count"] > 1)].head(19)

,count,max
order_id,,
0016dfedd97fc2950e388d2971d718c7,2,2
002f19a65a2ddd70a090297872e6d64e,2,2
0071ee2429bc1efdc43aa3e073a5290e,2,2
009ac365164f8e06f59d18a08045f6c4,6,6
00b4a910f64f24dbcac04fe54088a443,2,2
00bd50cdd31bd22e9081e6e2d5b3577b,3,3
00c405bd71187154a7846862f585a9d4,7,7
00c95282163553a982f38481f9488481,4,4
00e6bc6b166eb28b4502c1cad4457248,2,2


In [155]:
order_payments[["order_id", "payment_sequential"]].duplicated().sum()

np.int64(0)

### Payment Sequence

- `payment_sequential` behaves as a sequence within each `order_id`.
- Multi-payment orders have sequential values such as `1, 2, 3, ...`.
- No duplicate `order_id` + `payment_sequential` combinations were found.
- The sequence behavior appears consistent across the dataset.

In [156]:
order_payments["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

In [157]:
order_payments[order_payments["payment_type"] == "not_defined"]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
51280,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.0
57411,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.0
94427,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.0


### Payment Type

- Four standard payment types were observed: `credit_card`, `boleto`, `voucher`, and `debit_card`.
- 3 records have `payment_type = "not_defined"`.
- These records also have `payment_value = 0.0`.
- The correct payment type cannot be determined from the available data, so these records should be handled explicitly during cleaning rather than assigned an arbitrary category.

In [158]:
order_payments.groupby("payment_type")["payment_installments"].agg(["count", "min", "max", "nunique"])

,count,min,max,nunique
payment_type,,,,
boleto,19784,1,1,1
credit_card,76795,0,24,24
debit_card,1529,1,1,1
not_defined,3,1,1,1
voucher,5775,1,1,1


In [159]:
order_payments[(order_payments["payment_type"] == "credit_card") & (order_payments["payment_installments"] == 0)]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
46982,744bade1fcf9ff3f31d860ace076d422,2,credit_card,0,58.69
79014,1a57108394169c0b47d8f876acc9ba2d,2,credit_card,0,129.94


- 2 credit-card records have `payment_installments = 0` despite having nonzero payment values.
- This differs from the installment pattern observed in the other credit-card payments.
- The meaning of `0` is unclear from the data alone and should be verified before making any cleaning decision.

In [160]:
order_payments[order_payments["payment_value"] == 0]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
19922,8bcbe01d44d147f901cd3192671144db,4,voucher,1,0.0
36822,fa65dad1b0e818e3ccc5cb0e39231352,14,voucher,1,0.0
43744,6ccb433e00daae1283ccc956189c82ae,4,voucher,1,0.0
51280,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.0
57411,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.0
62674,45ed6e85398a87c253db47c2d9f48216,3,voucher,1,0.0
77885,fa65dad1b0e818e3ccc5cb0e39231352,13,voucher,1,0.0
94427,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.0
100766,b23878b3e8eb4d25a158f57d96331b18,4,voucher,1,0.0


In [161]:
order_payments[
    order_payments["order_id"] == "fa65dad1b0e818e3ccc5cb0e39231352"
][[
    "order_id",
    "payment_sequential",
    "payment_type",
    "payment_installments",
    "payment_value"
]].sort_values("payment_sequential")

,order_id,payment_sequential,payment_type,payment_installments,payment_value
14321,fa65dad1b0e818e3ccc5cb0e39231352,1,voucher,1,3.71
23074,fa65dad1b0e818e3ccc5cb0e39231352,2,voucher,1,8.51
65641,fa65dad1b0e818e3ccc5cb0e39231352,3,voucher,1,2.95
9985,fa65dad1b0e818e3ccc5cb0e39231352,4,voucher,1,29.16
28330,fa65dad1b0e818e3ccc5cb0e39231352,5,voucher,1,0.66
29648,fa65dad1b0e818e3ccc5cb0e39231352,6,voucher,1,5.02
82593,fa65dad1b0e818e3ccc5cb0e39231352,7,voucher,1,0.32
68853,fa65dad1b0e818e3ccc5cb0e39231352,8,voucher,1,26.02
17274,fa65dad1b0e818e3ccc5cb0e39231352,9,voucher,1,1.08
19565,fa65dad1b0e818e3ccc5cb0e39231352,10,voucher,1,12.86


### Payment Sequence

- `payment_sequential` behaves as a sequence within each `order_id`.
- Orders can contain multiple payment records, including multiple voucher payments.
- No duplicate `order_id` + `payment_sequential` combinations were found.
- `payment_sequential` represents the payment record's position within an order, not the number of installments.

### Payment Type

- `credit_card`, `boleto`, `voucher`, and `debit_card` are the main payment types.
- 3 records have `payment_type = "not_defined"`.
- These 3 records have `payment_value = 0.0`.
- Zero-value payments also occur with vouchers, so a zero `payment_value` is not inherently invalid.
- 2 credit-card records have `payment_installments = 0` despite having nonzero payment values. The meaning of this value is unclear and should be verified before cleaning.
- No negative `payment_value` values were found.

In [162]:
order_reviews = datasets[r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_order_reviews_dataset.csv"]

order_reviews.shape

(99224, 7)

In [163]:
order_reviews.describe()
order_reviews.columns

Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='str')

### Date and Timestamp Columns Stored as Strings

In the raw `order_reviews` dataset, the date and timestamp columns are stored as strings:

```text
review_creation_date      → str
review_answer_timestamp   → str

In [164]:
order_payments[(order_payments["payment_type"] == "credit_card") & (order_payments["payment_installments"] == 0)]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
46982,744bade1fcf9ff3f31d860ace076d422,2,credit_card,0,58.69
79014,1a57108394169c0b47d8f876acc9ba2d,2,credit_card,0,129.94


In [165]:
order_reviews["order_id"].isin(orders["order_id"]).value_counts()

order_id
True    99224
Name: count, dtype: int64

In [166]:
order_reviews["order_id"].value_counts().value_counts()

count
1    98126
2      543
3        4
Name: count, dtype: int64

### Review-to-Order Relationship

- `order_id` is not unique in `order_reviews`.
- Most orders have one review:
  - 98,126 orders → 1 review
  - 543 orders → 2 reviews
  - 4 orders → 3 reviews
- Therefore, `order_reviews` has a one-to-many relationship with `orders`.

Check:

```python
order_reviews["order_id"].value_counts().value_counts().sort_index()

In [167]:
order_reviews["review_id"].value_counts().value_counts()

count
1    97621
2      764
3       25
Name: count, dtype: int64

In [168]:
order_reviews[order_reviews["review_id"].duplicated(keep = False)].sort_values("review_id").head(20)

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
46678,00130cbe1f9d422698c812ed8ded1919,dfcdfc43867d1c1381bfaf62d6b9c195,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecid...",2018-03-07 00:00:00,2018-03-20 18:08:23
29841,00130cbe1f9d422698c812ed8ded1919,04a28263e085d399c97ae49e0b477efa,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecid...",2018-03-07 00:00:00,2018-03-20 18:08:23
90677,0115633a9c298b6a98bcbe4eee75345f,78a4201f58af3463bdab842eea4bc801,5,NaN,NaN,2017-09-21 00:00:00,2017-09-26 03:27:47
63193,0115633a9c298b6a98bcbe4eee75345f,0c9850b2c179c1ef60d2855e2751d1fa,5,NaN,NaN,2017-09-21 00:00:00,2017-09-26 03:27:47
92876,0174caf0ee5964646040cd94e15ac95e,f93a732712407c02dce5dd5088d0f47b,1,NaN,Produto entregue dentro de embalagem do fornec...,2018-03-07 00:00:00,2018-03-08 03:00:53
57280,0174caf0ee5964646040cd94e15ac95e,74db91e33b4e1fd865356c89a61abf1f,1,NaN,Produto entregue dentro de embalagem do fornec...,2018-03-07 00:00:00,2018-03-08 03:00:53
54832,017808d29fd1f942d97e50184dfb4c13,8daaa9e99d60fbba579cc1c3e3bfae01,5,NaN,NaN,2018-03-02 00:00:00,2018-03-05 01:43:30
99167,017808d29fd1f942d97e50184dfb4c13,b1461c8882153b5fe68307c46a506e39,5,NaN,NaN,2018-03-02 00:00:00,2018-03-05 01:43:30
20621,0254bd905dc677a6078990aad3331a36,5bf226cf882c5bf4247f89a97c86f273,1,NaN,O pedido consta de 2 produtos e até agora rece...,2017-09-09 00:00:00,2017-09-13 09:52:44
96080,0254bd905dc677a6078990aad3331a36,331b367bdd766f3d1cf518777317b5d9,1,NaN,O pedido consta de 2 produtos e até agora rece...,2017-09-09 00:00:00,2017-09-13 09:52:44


### Cleaning Decision: Duplicate Review IDs

Repeated `review_id` values were identified and inspected. Several repeated IDs contain the same review information but are associated with different `order_id` values.

These records are considered suspicious duplicate reviews.

They will not be removed during inspection. During cleaning, the duplicated review IDs will be investigated further before a retention rule is applied. I will not arbitrarily remove rows based only on `review_id`.

In [169]:
products.info()

<class 'pandas.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  str    
 1   product_category_name       32341 non-null  str    
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), str(2)
memory usage: 2.3 MB


### Column Name Consistency

- I found spelling errors in two column names:
  - `product_name_lenght`
  - `product_description_lenght`
- The intended spelling is `length`.
- The underlying values are not necessarily affected by this naming issue.
- I will correct the column names during the cleaning stage to make the dataset schema clear and consistent.

In [170]:
products.product_id.nunique()

32951

In [171]:
products["product_category_name"].isnull().value_counts()

product_category_name
False    32341
True       610
Name: count, dtype: int64

### Missing Product Information

- I found 610 products with missing `product_category_name`.
- These same 610 products also have missing:
  - `product_name_length`
  - `product_description_length`
  - `product_photos_qty`
- The physical product attributes such as weight and dimensions are still populated for these records.
- This suggests that the missing values are concentrated in product descriptive information rather than affecting the entire product record.
- I will investigate the impact of these missing values during the cleaning stage rather than assuming they are errors.

In [172]:
products[products["product_category_name"].isna()]["product_id"].isin(
    order_items["product_id"]
).value_counts()

product_id
True    610
Name: count, dtype: int64

In [173]:
products[products["product_weight_g"].isna()]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
8578,09ff539a621711667c43eba6a3bd8466,bebes,60.0,865.0,3.0,NaN,NaN,NaN,NaN
18851,5eb564652db742ff8f28759cd8d2652a,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Missing Physical Product Measurements

- I found 2 products with missing `product_weight_g`.
- These same 2 products also have missing `product_length_cm`, `product_height_cm`, and `product_width_cm`.
- One of the products has other descriptive information available, while the other has missing descriptive and physical information.
- I will not invent or estimate the missing measurements during cleaning without a reliable source for the values.

In [174]:
sellers = datasets[
    r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_sellers_dataset.csv"
]

sellers.info()

<class 'pandas.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   seller_id               3095 non-null   str  
 1   seller_zip_code_prefix  3095 non-null   int64
 2   seller_city             3095 non-null   str  
 3   seller_state            3095 non-null   str  
dtypes: int64(1), str(3)
memory usage: 96.8 KB


In [175]:
sellers.seller_id.nunique()

3095

In [176]:
order_items["seller_id"].isin(sellers["seller_id"]).value_counts()

seller_id
True    112650
Name: count, dtype: int64

In [177]:
sellers.groupby("seller_zip_code_prefix")["seller_state"].nunique().value_counts().sort_index()

seller_state
1    2229
2      16
3       1
Name: count, dtype: int64

In [178]:
seller_zip_state_check = (sellers.groupby("seller_zip_code_prefix")["seller_state"].nunique()
)

seller_zip_state_check[seller_zip_state_check > 1]

seller_zip_code_prefix
21210    2
22783    2
31160    2
36010    2
37540    2
37795    2
44600    2
80240    2
81020    2
81560    2
83020    2
83321    2
85960    3
88075    2
88301    2
89052    2
95076    2
Name: seller_state, dtype: int64

In [179]:
sellers[
    sellers["seller_zip_code_prefix"].isin(
        seller_zip_state_check[seller_zip_state_check > 1].index
    )
].sort_values("seller_zip_code_prefix")

,seller_id,seller_zip_code_prefix,seller_city,seller_state
114,392f7f2c797e4dc077e4311bde2ab8ce,21210,rio de janeiro,RN
1591,95e03ca3d4146e4011985981aeb959b9,21210,rio de janeiro,RJ
364,c716e0b86ed568878475b60fbb6323ad,22783,rio de janeiro,SP
299,b026643914f651e8371425e5b8a1e780,22783,rio de janeiro,RJ
1693,2c00c85d30361cd2ced2969cffbbffa3,31160,belo horizonte,MG
3036,6b9b80d53ba3676eafe60268a810b5a1,31160,belo horizonte,SP
2509,29e6e05b65f93bc224c644319dc52ecf,36010,juiz de fora,MG
680,d9d43faf741ceaa8da52fdbb81d9e0e3,36010,juiz de fora,SP
3012,f789d2c4f2c2eb38fc4373e7a4b35264,36010,juiz de fora,MG
1299,57df9869a600bd6b7c405f2a862eccfb,37540,santa rita do sapucai,MG


### Seller Geographic Consistency

- I checked whether `seller_zip_code_prefix` consistently maps to one `seller_state`.
- Most ZIP prefixes map to a single state, but 16 prefixes map to two states and 1 prefix maps to three states.
- I inspected the affected seller records and found several cases where the same city and ZIP prefix are associated with different states.
- Examples include `rio de janeiro` being associated with both `RN` and `RJ`, and `curitiba` being associated with both `PR` and `SP`.
- These records indicate potential inconsistencies in the seller geographic information.
- I will not manually correct the state values without a reliable reference source. I will revisit this issue during cleaning and decide whether the affected records should be corrected, standardized, or left unchanged.

In [180]:
geolocation = datasets[
    r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\olist_geolocation_dataset.csv"
]

geolocation.shape

(1000163, 5)

In [181]:
geolocation.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  str    
 4   geolocation_state            1000163 non-null  str    
dtypes: float64(2), int64(1), str(2)
memory usage: 38.2 MB


In [182]:
geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [183]:
geolocation["geolocation_zip_code_prefix"].value_counts().describe()


count    19015.000000
mean        52.598633
std         72.057907
min          1.000000
25%         10.000000
50%         29.000000
75%         66.500000
max       1146.000000
Name: count, dtype: float64

In [184]:
geolocation.duplicated().sum()

np.int64(261831)

In [185]:
geolocation.drop_duplicates().shape

(738332, 5)

### Exact Duplicate Geolocation Records

- I found 261,831 exact duplicate rows in the geolocation dataset.
- The dataset contains 1,000,163 rows in total.
- Removing exact duplicates would reduce the dataset to 738,332 rows.
- Repeated ZIP prefixes are expected because a ZIP prefix can have multiple geographic records.
- However, exact duplicate rows contain no additional geographic information and are redundant.
- I will remove exact duplicate rows during the cleaning stage.

In [186]:
geolocation[["geolocation_lat", "geolocation_lng"]].describe()

,geolocation_lat,geolocation_lng
count,1.000163e+06,1.000163e+06
mean,-2.117615e+01,-4.639054e+01
std,5.715866e+00,4.269748e+00
min,-3.660537e+01,-1.014668e+02
25%,-2.360355e+01,-4.857317e+01
50%,-2.291938e+01,-4.663788e+01
75%,-1.997962e+01,-4.376771e+01
max,4.506593e+01,1.211054e+02


In [187]:
geolocation[
    (geolocation["geolocation_lat"] > 6) |
    (geolocation["geolocation_lng"] > -34)
].shape

(37, 5)

### Geographic Coordinate Anomalies

- I inspected the latitude and longitude values using `describe()` to understand their ranges.
- I used broad geographic bounds for Brazil as an initial screening rule:
  - Latitude should generally fall between approximately -34° and +6°.
  - Longitude should generally fall between approximately -74° and -34°.
- I used these bounds as a way to identify obvious outliers, not as an exact definition of Brazil's borders.
- I flagged records where:
  - `geolocation_lat > 6`, or
  - `geolocation_lng > -34`

```python
out_of_range = geolocation[
    (geolocation["geolocation_lat"] > 6) |
    (geolocation["geolocation_lng"] > -34)
]

In [190]:
category_translation = datasets[
    r"C:\Users\Abdul\Documents\VS_code_projects\OLIST_Data_Cleaning\data\raw\product_category_name_translation.csv"
]
category_translation.shape

(71, 2)

In [191]:
category_translation.info()

<class 'pandas.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   product_category_name          71 non-null     str  
 1   product_category_name_english  71 non-null     str  
dtypes: str(2)
memory usage: 1.2 KB


In [192]:
category_translation.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [193]:
category_translation["product_category_name"].nunique()

71

In [196]:
products[
    products["product_category_name"].notna()
]["product_category_name"].isin(
   category_translation["product_category_name"]
).value_counts()

product_category_name
True     32328
False       13
Name: count, dtype: int64

In [200]:
product_categories = products.loc[
    products["product_category_name"].notna(),
    "product_category_name"
]

category_exists = product_categories.isin(
    category_translation["product_category_name"]
)

missing_categories = product_categories[~category_exists].unique()

missing_categories

<StringArray>
['pc_gamer', 'portateis_cozinha_e_preparadores_de_alimentos']
Length: 2, dtype: str

### Missing Category Translations

- I checked whether product categories with non-null values have a corresponding entry in the translation table.
- 32,328 product rows have categories that exist in the translation table.
- 13 product rows contain categories that do not have a translation.
- These 13 rows correspond to only 2 unique category names:
  - `pc_gamer`
  - `portateis_cozinha_e_preparadores_de_alimentos`
- I will not invent English translations for these categories.
- I will investigate whether these categories can be reliably translated during the cleaning stage.

### ----------------------------------------------------------------------------------

### ----------------------------------------------------------------------------------

# Raw Data Inspection: Conclusion

I completed the initial inspection of all 9 Olist CSV datasets.

The purpose of this stage was to understand the structure of the raw data, identify meaningful data-quality issues, and determine which issues require attention during cleaning.

I followed the process:

Find
↓
Measure
↓
Inspect
↓
Decide
↓
Move on

I did not modify or clean the raw data during this stage.

## 1. Customers

- I found no missing values.
- `customer_id` is unique across all 99,441 records.
- `customer_unique_id` is not unique, meaning the same underlying customer can appear in multiple customer records.
- Every `customer_id` in `orders` exists in `customers`.
- Every customer ZIP prefix maps to one state.
- I found 39 ZIP prefixes associated with multiple city names, affecting 476 customer records.
- I consider these potential city/ZIP consistency issues and will not investigate them further without a clear reason.

## 2. Orders

- The dataset contains 99,441 orders and 8 columns.
- `order_id` is unique.
- `customer_id` is unique within `orders`.
- Every `customer_id` in `orders` exists in `customers`.
- The date and timestamp columns were initially stored as strings.
- Missing values occur in several order date columns.
- Most missing delivery dates are associated with orders that were not delivered.
- 8 orders have `order_status = "delivered"` but have no customer delivery date.
- I tested the date columns with `pd.to_datetime()` and found no additional malformed dates beyond the existing missing values.
- I will convert the date and timestamp columns to datetime during cleaning.
- I will investigate the 8 delivered orders with missing delivery dates during cleaning.

## 3. Order Items

- The dataset contains 112,650 order-item records.
- I found no missing values.
- `order_id` is not unique because an order can contain multiple items.
- `order_item_id` is a sequence within each order rather than a globally unique identifier.
- The meaningful key is the combination of `order_id` and `order_item_id`.
- I found no duplicate `order_id` + `order_item_id` combinations.
- The item sequence was consistent across orders.
- Every `order_id` exists in `orders`.
- Every `product_id` exists in `products`.
- I did not identify a significant data-quality issue in this table during inspection.

## 4. Order Payments

- The dataset contains 103,886 payment records.
- I found no missing values.
- `payment_sequential` behaves as a sequence within each `order_id`.
- I found no duplicate `order_id` + `payment_sequential` combinations.
- Multiple payment records can legitimately belong to the same order.
- `payment_type = "not_defined"` occurs in 3 records.
- These 3 records have a payment value of 0.0.
- I found that zero-value payments also occur with vouchers, so I will not automatically treat zero payment values as errors.
- Credit-card payments are the only payment type with varying installment counts.
- 2 credit-card records have `payment_installments = 0` despite having nonzero payment values.
- I do not know whether those 0 values are valid, so I will verify their meaning before making a cleaning decision.
- I found no negative payment values.
- High `payment_sequential` values are not automatically errors because some orders legitimately contain many payment records.

## 5. Order Reviews

- The dataset contains 99,224 review records and 7 columns.
- `review_score` is the only numeric field.
- The review date and timestamp columns are stored as strings.
- `review_comment_title` and `review_comment_message` contain many missing values.
- I will preserve customer review text as originally written rather than modifying the wording.
- I will only change the data type of the date and timestamp fields during cleaning.
- Every `order_id` in `order_reviews` exists in `orders`.
- `review_id` is not unique.
- 97,621 review IDs occur once, 764 occur twice, and 25 occur three times.
- I inspected repeated review IDs and found cases where the same review information is associated with different order IDs.
- These records appear suspicious and may represent duplicated review records.
- I will investigate them further during cleaning rather than arbitrarily deleting rows based only on `review_id`.

## 6. Products

- The dataset contains 32,951 products and 9 columns.
- `product_id` is unique.
- I found 610 products with missing `product_category_name`.
- These same products also have missing product name length, description length, and photo count.
- All 610 products appear in `order_items`, meaning the missing descriptive information affects products that occur in actual orders.
- I found 2 products with missing physical measurements.
- The missing physical measurements occur across weight, length, height, and width for these products.
- Two column names contain spelling errors:
  - `product_name_lenght`
  - `product_description_lenght`
- I will correct these column names during cleaning.
- I will not invent missing product categories or physical measurements without a reliable source.

## 7. Sellers

- The dataset contains 3,095 sellers and 4 columns.
- I found no missing values.
- `seller_id` is unique.
- Every `seller_id` referenced by `order_items` exists in `sellers`.
- I found ZIP prefixes associated with multiple states.
- 16 ZIP prefixes occur across 2 states, and 1 ZIP prefix occurs across 3 states.
- After inspecting the affected records, I found examples where the same city and ZIP prefix are associated with different states.
- These indicate potential inconsistencies in the seller geographic information.
- I will not manually correct the state values without a reliable reference source.

## 8. Geolocation

- The dataset contains 1,000,163 records and 5 columns.
- I found no missing values.
- I found 19,015 unique ZIP code prefixes.
- A ZIP prefix can have many geolocation records, so repeated ZIP prefixes are not automatically duplicates.
- I found 261,831 exact duplicate rows.
- Removing exact duplicate rows would reduce the dataset from 1,000,163 to 738,332 records.
- Exact duplicate rows contain no additional geographic information, so I will remove them during cleaning.
- I inspected latitude and longitude ranges using broad geographic bounds for Brazil as a screening method.
- I used approximate bounds of -34° to +6° latitude and -74° to -34° longitude to identify obvious geographic outliers.
- These bounds were used as a screening tool rather than an exact definition of Brazil's borders.
- I identified 37 records with coordinates outside these broad bounds.
- I inspected examples and found inconsistencies between the coordinates and their associated city/state information.
- I will not modify those coordinates without a reliable reference source.
- I will document them as potential geographic data-quality issues and leave them unchanged unless reliable evidence becomes available.

## 9. Product Category Name Translation

- The dataset contains 71 rows and 2 columns.
- I found no missing values.
- Both columns are stored as strings.
- `product_category_name` is unique across all 71 rows.
- I checked whether product categories appearing in `products` have a corresponding translation.
- 32,328 product rows contain categories that exist in the translation table.
- 13 product rows contain categories that do not have a translation.
- These 13 rows correspond to only 2 unique category names:
  - `pc_gamer`
  - `portateis_cozinha_e_preparadores_de_alimentos`
- I will not invent English translations for these categories.
- I will investigate whether reliable translations can be established during the cleaning stage.

## Main Data-Quality Findings

| Dataset | Finding | Planned action |
|---|---|---|
| Customers | 39 ZIP prefixes have multiple city names | Document and evaluate if necessary |
| Orders | Date columns stored as strings | Convert to datetime |
| Orders | 8 delivered orders missing delivery date | Investigate |
| Order Payments | 2 credit-card records have 0 installments with nonzero payment values | Verify before changing |
| Order Reviews | Repeated `review_id` values | Investigate before removing |
| Order Reviews | Date/timestamp columns stored as strings | Convert to datetime |
| Products | 610 products missing descriptive information | Investigate handling |
| Products | 2 products missing physical measurements | Investigate handling |
| Products | 2 misspelled column names | Rename during cleaning |
| Sellers | Geographic inconsistencies | Do not manually correct without reliable source |
| Geolocation | 261,831 exact duplicate rows | Remove exact duplicates |
| Geolocation | 37 suspicious coordinate records | Document; do not modify without evidence |
| Category Translation | 2 categories have no translation | Investigate before filling |

## Inspection Conclusion

I now have a clear understanding of the structure and major quality issues across the Olist dataset.

I found that not every unusual value is an error. Some repeated values are legitimate because of one-to-many relationships between tables, while some missing values are expected based on the state of an order.

I also found several issues that require cleaning or further investigation, including:

- inconsistent data types
- missing values
- suspicious duplicate review records
- misspelled column names
- exact duplicate geolocation records
- geographic inconsistencies
- missing category translations
- potentially inconsistent payment installment values

I will not modify the raw datasets directly.

The project will now move from inspection into the cleaning stage.